### ISSUE TITLE
Verify Pulumi SDKs work with .NET6

### ISSUE THREAD

**Issue by emiliza** [pulumi/pulumi#8153]
Part of: https://github.com/pulumi/pulumi/issues/7911

.NET6 will be released on Nov 9. We should at a minimum verify our SDKs are compatible with the latest version 

Results:
You can build our sdks using the new  `dotnet 6.0.100-rc.2.21505.57` CLI tool (targeting `netcoreapp3.1` as before) and everything seems to work (all tests green, sample projects spun up).  All .NET unit tests pass. All integration tests pass. 

I have successfully ran `pulumi up` for projects targeting `net6.0` (and integrating with the original SDKs) for
|   | macOS | Windows10 |
|-|-|-|
| `pulumi new aws-csharp` |  ✅  |  ✅  |
| Automation AP (Inline Program) | ✅ | ✅ |
|  Binary |✅ | ✅| 

I am assuming that linux will behave similarly to macOS.

**Comment by mikhailshilkov** [comment]
I'd say the main scenario we should validate is that an end user can set the target to net6.0 for their Pulumi project and use our existing SDKs with no issues.

**Comment by atrauzzi** [comment]
That would be great, I'm really hoping to jump into C# 10.

**Comment by gitfool** [comment]
@mikhailshilkov that's effectively what I did in [test commit](https://github.com/gitfool/Pulumi.Dungeon/commit/09389381eb0ae7f82789b865df8120a30d6be6b6) above, so I hope single file executables are covered.

**Comment by iwahbe** [comment]
@gitfool , can I get the results of `pulumi about`. I've tried running (Pulumi.Dungeon)[https://github.com/gitfool/Pulumi.Dungeon] on both Mac and windows and neither give me the error you showed.

**Comment by gitfool** [comment]
@iwahbe here you go:
```
user@32f2c6035bc0:~/dungeon$ pulumi about
Logging in using access token from PULUMI_ACCESS_TOKEN
CLI
Version      3.16.0
Go Version   go1.17.2
Go Compiler  gc

Host
OS       ubuntu
Version  20.04
Arch     x86_64

Backend
Name     pulumi.com
URL      https://app.pulumi.com/pharos-sfausett
User     pharos-sfausett

Pulumi locates its logs in /tmp by default
warning: Failed to get information about the plugin: no Pulumi.yaml project file found (searching upwards from /home/user/dungeon). If you have not created a project yet, use `pulumi new` to do so
warning: Failed to read project: no Pulumi.yaml project file found (searching upwards from /home/user/dungeon). If you have not created a project yet, use `pulumi new` to do so
warning: Failed to get information about the current stack: no Pulumi.yaml project file found (searching upwards from /home/user/dungeon). If you have not created a project yet, use `pulumi new` to do so
```
I probably didn't give you enough breadcrumbs to actually run it; did you do any of the following?
* Checkout the [test-net6](https://github.com/gitfool/Pulumi.Dungeon/tree/test-net6) branch
* Edit [config/test.yaml](https://github.com/gitfool/Pulumi.Dungeon/blob/test-net6/Cli/config/test.yaml) to override aws and pulumi config with your accounts etc
* Run the [cake build](https://github.com/cake-build/cake) to generate the docker image with single file executable and config
  * `dotnet tool restore && dotnet cake --verbosity=verbose`
* Run the docker image passing relevant env vars
  * `docker run -it --rm --entrypoint bash -e AWS_ACCESS_KEY_ID -e AWS_SECRET_ACCESS_KEY -e AWS_REGION -e PULUMI_ACCESS_TOKEN dockfool/pulumi-dungeon --login`
* Run the deploy
  * `d deploy test --skip-preview --yes`

**Comment by iwahbe** [comment]
I got to bullets 1, 2 and then ran the project locally. This is extremely helpful.

**Comment by gitfool** [comment]
I haven't tried running on macOS or Windows10 as a single file executable, but I presume you're doing so by building a release build after updating the [RuntimeIdentifier](https://github.com/gitfool/Pulumi.Dungeon/blob/test-net6/Cli/Cli.csproj#L22) accordingly?

Note: When developing in Windows I don't generate a single file executable, hence the debug build does not do this. Ultimately it's the linux build that matters, which does generate a single file executable, as that's what is run for CI/CD.

**Comment by iwahbe** [comment]
No, I was using debug mode. Pulumi has a concept of a single file binary, which is different from .NET's concept of a single file binary. That was my bad. I have updated my summary to reflect that I tested a Pulumi binary project. Pulumi.Dungeon fails on all platforms in release mode with the same error. Thanks for confirming.

> The failure looks due to a change in linker trimming behaviour between net5.0 and net6.0. 

I do get a successful run with `<PublishTrimmed>false</PublishTrimmed>`, so I suspect you are right.
